internal/graph/check_fast_path.go

```diff
@@ -0,0 +1,674 @@
+
+	rightSet := hashset.New()
+	leftSet := hashset.New()
+
+	r, ok := <-rightChan
+	if !ok {
+		return res, nil
+	}
+	if r.err != nil {
+		return nil, r.err
+	}
+	rightSet.Add(r.userset)
+	for leftOpen || rightOpen {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case msg, ok := <-leftChan:
+			if !ok {
+				leftOpen = false
+				if leftSet.Size() == 0 {
+					return res, nil
+				}
+				break
+			}
+			if msg.err != nil {
+				return nil, msg.err
+			}
+			for {
+				t, err := msg.iter.Next(ctx)
+				if err != nil {
+					msg.iter.Stop()
+					if storage.IterIsDoneOrCancelled(err) {
+						break
+					}
+					return nil, err
+				}
+				if processUsersetMessage(t.GetObject(), leftSet, rightSet) {
+					msg.iter.Stop()
+					res.Allowed = true
+					return res, nil
+				}
+			}
+		case msg, ok := <-rightChan:
+			if !ok {
+				rightOpen = false
+				break
+			}
+			if msg.err != nil {
+				return nil, msg.err
+			}
+			if processUsersetMessage(msg.userset, rightSet, leftSet) {
+				res.Allowed = true
+				return res, nil
+			}
+		}
+	}
+	return res, nil
+}
+
+func (c *LocalChecker) checkTTUFastPathV2(ctx context.Context, req *ResolveCheckRequest, rewrite *openfgav1.Userset, iter storage.TupleKeyIterator) (*ResolveCheckResponse, error) {
+	ctx, span := tracer.Start(ctx, "checkTTUFastPathV2")
+	defer span.End()
+	typesys, _ := typesystem.TypesystemFromContext(ctx)
+	objectType := tuple.GetType(req.GetTupleKey().GetObject())
+	tuplesetRelation := rewrite.GetTupleToUserset().GetTupleset().GetRelation()
+	computedRelation := rewrite.GetTupleToUserset().GetComputedUserset().GetRelation()
+
+	possibleParents, err := typesys.GetDirectlyRelatedUserTypes(objectType, tuplesetRelation)
+	if err != nil {
+		return nil, err
+	}
+
+	cancellableCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	leftChans := make([]chan *iteratorMsg, 0)
+	for _, parentType := range possibleParents {
+		r := req.clone()
+		r.TupleKey = &openfgav1.TupleKey{
+			Object:   tuple.BuildObject(parentType.GetType(), "ignore"),
+			Relation: computedRelation,
+			User:     r.GetTupleKey().GetUser(),
+		}
+		rel, err := typesys.GetRelation(parentType.GetType(), computedRelation)
+		if err != nil {
+			// NOTE: is there a better way to check and filter rather than skipping?
+			// other paths can be reachable
+			continue
+		}
+		leftChan, err := c.fastPathRewrite(cancellableCtx, r, rel.GetRewrite())
+		if err != nil {
+			return nil, err
+		}
+		leftChans = append(leftChans, leftChan)
+	}
+
+	if len(leftChans) == 0 {
+		// NOTE: this should be an error right?
+		return &ResolveCheckResponse{
+			Allowed: false,
+		}, nil
+	}
+
+	return c.resolveFastPath(ctx, leftChans, wrapIterator(TTUKind, iter))
+}
+
+// NOTE: Can we make this generic and move it to concurrency pkg?
+func fanInIteratorChannels(ctx context.Context, chans []chan *iteratorMsg) chan *iteratorMsg {
+	limit := len(chans)
+	pool := concurrency.NewPool(ctx, limit)
+	out := make(chan *iteratorMsg, limit)
```

Comment:
You put a limit of `len(chan)`, however, there is a double `for` loop down below, which means that potentially we could be trying to send _more_ than `len(chan)` messages into it. Should be make the channel size bigger to accomodate for this?

Reply:
> Should be make the channel size bigger to accomodate for this?

Nop, having them block for a while will allow us to have backpressure control. If the consumer is overwhelmed it should signal the producer to backoff. Trying to prevent the N*M problem we have in the normal path for Check.